test/compilable/must_use_not_reserved.d

```diff
@@ -0,0 +1,5 @@
+import core.attribute;
+
+@mustUse int n;
+@mustUse alias A = int;
```

Comment:
Shouldn't this be a warning / error? Seems like an error that is very easy to do and silently allows broken code.

Reply:
This is correct per the formal spec in DIP 1038, and is consistent with how other attributes work (e.g., `@nogc`).

Reply:
> This is correct per the formal spec in DIP 1038, 

The  "Diagnostics" section (and the DIP in general) doesn't define rules for `alias` declarations - unless the absence is to be interpreted as "no diagnostics" instead of "not considered".

> and is consistent with how other attributes work (e.g., `@nogc`).

That's only partially true. E.g. the `@nogc` attribute is discarded but misuse usually results in a subsequent error because `@nogc` code tries to call a `non-@nogc` function pointer / delegate.

Reply:
The relevant rule, from the "Formal Specification" section, is:

> It is a compile-time error to discard an expression if all of the following are true:
> [...]
> * its type is a struct or union type whose declaration is annotated with `@mustUse`.

Whether that type is referred to via an `alias` or directly, it is the attribute on the type itself that matters.

Reply:
Indeed but that also assumes knowledge of the wording in the DIP. The current behavior might not be as obvious for common users.

Whether `@mustUse alias ...` should raise  additional diagnostics is not defined / forbidden by the DIP. The current implementation already raises errors when annotating e.g. functions with `@mustUse` for the same reason - it's not supported and would have no effect.

Reply:
> Indeed but that also assumes knowledge of the wording in the DIP.

The wording from the DIP's "Formal Specification" section will be incorporated into the language spec.

> Whether `@mustUse alias ...` should raise additional diagnostics is not defined / forbidden by the DIP.

The compiler is, of course, free to issue additional diagnostics above and beyond what the spec requires.

As a general rule, DMD does not currently issue diagnostics for attributes used on declarations that do not support them. If we wish to revisit that policy, fine, but I do not think this PR is the place for it.

> The current implementation already raises errors when annotating e.g. functions with `@mustUse` for the same reason - it's not supported and would have no effect.

`@mustUse` on functions is specifically *reserved* by the DIP for potential future expansion. `@mustUse` on aliases is not reserved.

Reply:
I agree with @pbackus. The DIP specifically states: "It is a compile-time error to annotate any function declaration or aggregate declaration other than a struct or union declaration with the @mustUse attribute. The purpose of this rule is to reserve such usage for possible future expansion.". So the implementation correctly covers the spec.

On the topic of ignoring attributes, this comes up once in a while. The general consensus is that we should ignore them because you also have the possibility to use attributes on-masse (e.g`@safe:`). If you are issue an error for all the declarations for which the attribute doesn't make sense then you are basically making en-masse attributes unusable.